### PR TITLE
Cleanup core thread error handling

### PR DESCRIPTION
**Pull request by joshuacolvin0**
* Fix message delivery retry
* If message delivery fails too many times, treat as fatal
* Add additional logging

### CODE REVIEW

File: packages/arb-avm-cpp/data_storage/src/arbcore.cpp [hunk 1]
```diff
@@ -1529,11 +1531,11 @@ bool ArbCore::threadBody(ThreadDataStruct& thread_data) {
             auto add_status = addMessages(message_data, thread_data.cache);
             if (!add_status.status.ok()) {
                 thread_data.add_messages_failure_count++;
-                message_data_error_string = add_status.status.ToString();
-                message_data_status = MESSAGES_ERROR;
```
Comment (hkalodner) [inline packages/arb-avm-cpp/data_storage/src/arbcore.cpp:1533]: Is the theory that this was what caused the error?

**Review by hkalodner — CHANGES_REQUESTED** [review]

File: packages/arb-avm-cpp/data_storage/src/arbcore.cpp [hunk 2]
```diff
@@ -1542,7 +1544,7 @@ bool ArbCore::threadBody(ThreadDataStruct& thread_data) {
                     return false;
                 }
 
-                std::cerr << "ArbCore addMessages non-fatal error "
+                std::cerr << "ArbCore addMessages non-fatal error number "
                           << thread_data.add_messages_failure_count << ": "
                           << message_data_error_string << "\n";
```
[the remark this answers is not in the document]
Comment (joshuacolvin0) [reply packages/arb-avm-cpp/data_storage/src/arbcore.cpp:1549]: :+1:

File: packages/arb-util/common/errors.go [hunk 3]
```diff
@@ -0,0 +1,12 @@
+package common
+
+import "strings"
+
+func IsFatalError(err error) bool {
+	if strings.Contains(err.Error(), "arbcore thread aborted") ||
```
[the remark this answers is not in the document]
Comment (joshuacolvin0) [reply packages/arb-util/common/errors.go:6]: :+1:

File: packages/arb-avm-cpp/cavm/carbcore.cpp [hunk 4]
```diff
@@ -92,13 +92,19 @@ int arbCoreDeliverMessages(CArbCore* arbcore_ptr,
...
+    auto item_count = sequencer_batch_items.size();
+    auto delayed_count = delayed_messages.size();
     try {
         auto status = arb_core->deliverMessages(
             previous_message_count, previous_inbox_acc,
             std::move(sequencer_batch_items), std::move(delayed_messages),
             reorg_message_count);
         return status;
     } catch (const std::exception& e) {
+        std::cerr << "exception while delivering messages, previous count: "
```
[the remark this answers is not in the document]
Comment (joshuacolvin0) [reply packages/arb-avm-cpp/cavm/carbcore.cpp:104]: :+1:

File: packages/arb-avm-cpp/data_storage/src/arbcore.cpp (hunk 1 above)
Comment (joshuacolvin0) [reply packages/arb-avm-cpp/data_storage/src/arbcore.cpp:1533]: It is definitely an error, not sure if it was the one we were hoping to fix

**Review by hkalodner — APPROVED** [review]
LGTM

**Pull request merged**